### ISSUE TITLE
Enhancement: Log a warning if configuration options aren't provided as expected types?

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#6403]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Spinning out from https://github.com/typescript-eslint/typescript-eslint/pull/6367#discussion_r1086065981: we have logic in `typescript-estree` to ensure provided configuration options are the expected types. But we don't show any indication about incorrect types to users who configure us incorrectly.

### Fail

```typescript
{
  parserOptions: {
    project: 123
  }
}
```

### Pass

```typescript
{
  parserOptions: {
    project: ['tsconfig.eslint.json']
  }
}
```

### Additional Info

Proposal:
* `console.warn` in v6?
* `throw` an error in v7?

**Comment by JoshuaKGoldberg** [comment]
ping @typescript-eslint/triage-team - any thoughts on this?

**Comment by bradzacher** [comment]
We should just throw? Logging is bad because people don't look at them and don't action them.

**Comment by JoshuaKGoldberg** [comment]
👍 . Adding to the v6 milestone, but we can always push to v6 if this ends up taking too long.

**Comment by Josh-Cena** [comment]
Is this a breaking change? Won't it just cause undefined behavior downstream anyway?

**Comment by JoshuaKGoldberg** [comment]
Yeah agreed the warning log is a much lighter breaking change than most. But I'd still think it qualifies. If something was previously passing some incorrect or unknown property, this will cause it to newly spit out output. I think that could break tooling if it was expecting things in a particular format.

**Comment by bradzacher** [comment]
When we log we have to ensure its to stderr. A non-trivial number of consumers assume that eslint output comes in certain formats and stdout logs can break that.

**Comment by Zamiell** [comment]
In there interim, is there some solution that end-users can use to validate their `parserOptions` at type-time?

As a quick example, my config file looks something like this (simplified):

```js
/** @type {import("eslint").Linter.Config} */
const config = {
  parserOptions: {
    sourceType: "module",
    ecmaVersion: "latest",
    tsconfigRootDir: REPO_ROOT,
    project: true,
  },

  ignorePatterns: ["**/dist/**"],
};
```

But I want to import the correct type, so that it looks something like this:

```
/** @type {import("eslint").Linter.Config} */
const config = {
  /** @type {import("@typescript-eslint/eslint-plugin").Linter.Config.ParserOptions} */
  parserOptions: {
    sourceType: "module",
    ecmaVersion: "latest",
    tsconfigRootDir: REPO_ROOT,
    project: true,
  },

  ignorePatterns: ["**/dist/**"],
};
```

But what is the correct import line?

**Comment by bradzacher** [comment]
@Zamiell nope - hard crashes at runtime.
The new flat config work would provide a typed function for people that might help but still not entirely likely to help as we need to support untyped keys for other parsers..

**Comment by JoshuaKGoldberg** [comment]
We talked internally: v8 has a lot of changes in it and nobody has taken ownership of this non-trivial change. Removing it from the v8 milestone.  We still want this, but it'll have to come in a future major version.
